code/datums/gamemode/powers/powers.dm

```diff
@@ -0,0 +1,372 @@
+/datum/power			
+	var/name = "Power"
+	var/desc = "Placeholder"
+	var/helptext = ""
+	var/buytext = ""
+	var/passive = FALSE     //is this a spell or a passive effect?
+	var/spellpath           //Path to a verb that contains the effects.
+	var/cost                //the cost of this power
+	var/datum/role/role 
+	var/obj/abstract/screen/movable/spell_master/spellmaster
+
+	var/store_in_memory = FALSE
+
+//basic proc for limiting powers in monkeys or whatever, override this with checks
+/datum/power/proc/can_use(var/mob/user)
+	return TRUE
+
+//adding the powers to the current ones, override this with passive effects
+/datum/power/proc/add_power(var/datum/role/R)
+	if (!istype(R) || !R)
+		return FALSE
+	if(is_type_in_list(src, R.current_powers))
```

Comment:
This coupled with `R.current_powers += src` seems questionable. Won't this result in `is_type_in_list` returning FALSE the second time this is run? That's because `is_type_in_list` will turn the list into an assoc list, and check whether `list[element] == 1`. `R.current_powers += src` is the equivalent of `R.current_powers[src] = null` instead.

Reply:
That would explain why current_powers was being longer/shorter than it should be when I viewed it in game. Will replacing ```R.current_powers += src``` with ```R.current_powers.Add(src)``` do the trick since I only want to add the power to the end of the list instead of creating a new index for it.

Reply:
`locate(type) in R.current_powers` and `R.current_powers += src` should work. 
`R.current_powers += src` and `R.current_powers.Add(src)` are equivalent, for the record.

Reply:
you didn't change this